### PR TITLE
Refactor replay mechanism and introduce `response` callback

### PR DESCRIPTION
**Pull request by kamalaknn**
Pending

 - [x] Docs, inline and Readme
 - [x] Tests for `response` callback

### CODE REVIEW

File: lib/runner/extensions/request-wrapper.command.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+var _ = require('lodash');
+
+module.exports = {
+    init: function (done) {
+        done();
+    },
+
+    triggers: ['response'],
+
+    process: {
```
Comment (shamasis) [inline lib/runner/extensions/request-wrapper.command.js]: rename this to `request` and current `request` to `http-request`

File: lib/runner/extensions/request-wrapper.command.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        requestWrapper: function (payload, next) {
+            var abortOnError = _.has(payload, 'abortOnError') ? payload.abortOnError : this.options.abortOnError,
+                complete = function (nextPayload, err) {
+                    this.triggers.response(
```
Comment (shamasis) [inline lib/runner/extensions/request-wrapper.command.js]: resolve variables here

File: lib/runner/extensions/request-wrapper.command.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+                        nextPayload.coords,
+                        nextPayload.response,
+                        nextPayload.request,
+                        payload.item,
+                        nextPayload.cookies
+                    );
+                    next(err && abortOnError ? err : null, nextPayload, err);
+                }.bind(this);
+
+            this.immediate('request', payload)
```
Comment (shamasis) [inline lib/runner/extensions/request-wrapper.command.js]: why not just queue?

File: lib/runner/extensions/request.command.js [hunk 4]
```diff
@@ -190,9 +205,10 @@ module.exports = {
                         context.response = res;
 
                         // run the post request helpers, which need to use the response, assigned above
-                        async.applyEachSeries(posthelpers, context, self, function (/* err */) {
+                        async.applyEachSeries(posthelpers, context, self, function (err, options) {
+                            if (err) { /** */ }
                             // @todo: need to handle err
-                            complete();
+                            complete(options[0]);
```
Comment (shamasis) [inline lib/runner/extensions/request.command.js]: need to explain what this is.

File: lib/runner/extensions/request.command.js [hunk 5]
```diff
@@ -190,9 +205,10 @@ module.exports = {
                         context.response = res;
 
                         // run the post request helpers, which need to use the response, assigned above
-                        async.applyEachSeries(posthelpers, context, self, function (/* err */) {
+                        async.applyEachSeries(posthelpers, context, self, function (err, options) {
+                            if (err) { /** */ }
```
Comment (shamasis) [inline lib/runner/extensions/request.command.js]: handle error here? question is, what nature of error will bubble up here and then are they supposed to stop run?

Comment (kamalaknn) [reply lib/runner/extensions/request.command.js]: Handled error and enabled test cases for asserting the error.

File: lib/runner/extensions/request-wrapper.command.js (hunk 3 above)
Comment (kamalaknn) [reply lib/runner/extensions/request-wrapper.command.js]: Added docs to explain why.

File: lib/runner/extensions/request-wrapper.command.js (hunk 2 above)
Comment (kamalaknn) [reply lib/runner/extensions/request-wrapper.command.js]: Not trivial to include in this PR. Will do it separately.

File: lib/runner/replay-controller.js [hunk 6]
```diff
@@ -55,45 +54,33 @@ _.assign(ReplayController.prototype, /** @lends ReplayController.prototype */{
...
-        // or schedule a request and continue with current queue
-        else {
-            this.run.interrupt('request', payload);
-            success();
-        }
+        this.run.immediate('httpRequest', payload)
+            .done(function (response) {
+                success(null, response);
+            })
+            .catch(function (err) {
```
Comment (kunagpal) [inline lib/runner/replay-controller.js]: Shouldn't this be a direct reference to `failure`?

File: lib/runner/request-helpers-postsend.js [hunk 7]
```diff
@@ -23,23 +24,17 @@ module.exports = [
...
-            // Auth wants to replay the request
-            var replayController = new ReplayController(context.replayState, run);
-
-            // finish up current request and schedule a replay
-            replayController.requestReplay(context, context.item, false, function (err) {
-                done(err);
-            }, function (err) {
-                done(err);
-            });
+            if (success) { return done(); }
```
Comment (kunagpal) [inline lib/runner/request-helpers-postsend.js]: Shouldn't this be an explicit `replay: false`?

File: lib/runner/replay-controller.js (hunk 6 above)
Comment (kamalaknn) [reply lib/runner/replay-controller.js]: If you are talking about `success` vs `failure`, here `success` is when `ReplayController` decides to send a request. `failure` is when it decided not to replay the request due to max retries exceeding and not the failure in the actual request.

File: lib/runner/request-helpers-postsend.js (hunk 7 above)
Comment (kamalaknn) [reply lib/runner/request-helpers-postsend.js]: IMHO, the replays should be an opt-in flow. I am not sure about forcing every other flow to turn off the. The way I see it, I want more options to be introduced. In that case wouldn't it be too much of an overhead?

**Comment by kamalaknn** [comment]
@kunagpal Overall this is not a breaking change. We are introducing a new callback without touching existing ones. The tests concur.

The only change to existing flows (as can be seen from changes in tests) is, errors in `post helpers` are now bubbled up. This I think needed to be fixed.

File: lib/runner/replay-controller.js (hunk 6 above)
Comment (kunagpal) [reply lib/runner/replay-controller.js]: Either way, `success(err);` looks weird.

File: lib/runner/extensions/event.command.js [hunk 8]
```diff
@@ -258,7 +258,7 @@ module.exports = {
 
                         cursorWithSource._source = 'script'; // @todo - get script type from the sandbox
 
-                        this.immediate('request', {
+                        this.immediate('httpRequest', {
```
Comment (shamasis) [inline lib/runner/extensions/event.command.js]: case insensitive - all lowercase

**Pull request merged**